File: tools/verify/render.c
/* Render tools/verify/cases.tsv through the C core and print a digest per case.
 * tools/verify/render.js prints the same digest via the legacy JS. If the two
 * outputs differ, the port is not faithful. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../../core/deck.h"
#include "../../core/font.h"

#define W 192
#define H 48

static uint32_t fnv1a(const uint8_t *p, size_t n) {
  uint32_t h = 2166136261u;
  for (size_t i = 0; i < n; i++) { h ^= p[i]; h *= 16777619u; }
  return h;
}

int main(int argc, char **argv) {
  const char *path = argc > 1 ? argv[1] : "tools/verify/cases.tsv";
  FILE *f = fopen(path, "r");
  if (!f) { perror(path); return 1; }

  static uint8_t px[W * H];
  const deck_geom_t geom = { W, H, DECK_LEVELS, 0 };
  deck_fb_t fb = { &geom, px };
  deck_clear(&fb);

  char line[1024];
  while (fgets(line, sizeof line, f)) {
    if (line[0] == '/' || line[0] == '\n') continue;
    size_t len = strlen(line);
    while (len && (line[len - 1] == '\n' || line[len - 1] == '\r')) line[--len] = 0;

    /* six tab-separated fields, then the text runs to end of line */
    char *fld[6];
    char *p = line;
    int ok = 1;
    for (int i = 0; i < 6; i++) {
      char *t = strchr(p, '\t');
      if (!t) { ok = 0; break; }
      *t = 0; fld[i] = p; p = t + 1;
    }
    if (!ok) continue;

    const char *name = fld[0];
    int fn = atoi(fld[1]), x = atoi(fld[2]), y = atoi(fld[3]);
    int inten = atoi(fld[4]), scale = atoi(fld[5]);
    const char *text = p;

    if (name[0] == '+') name++; else deck_clear(&fb);

    int adv = (fn == 3) ? deck_text3(&fb, x, y, text, (uint8_t)inten)
                        : deck_text5(&fb, x, y, text, (uint8_t)inten, scale);
    int width = (fn == 3) ? deck_width3(text) : deck_width5(text, scale);

    int nz = 0, x0 = W, y0 = H, x1 = -1, y1 = -1;
    for (int yy = 0; yy < H; yy++)
      for (int xx = 0; xx < W; xx++)
        if (px[yy * W + xx]) {
          nz++;
          if (xx < x0) x0 = xx;
          if (xx > x1) x1 = xx;
          if (yy < y0) y0 = yy;
          if (yy > y1) y1 = yy;
        }

    printf("%-16s hash=%08x nz=%-5d bbox=%d,%d,%d,%d adv=%d width=%d\n",
           name, fnv1a(px, sizeof px), nz, x0, y0, x1, y1, adv, width);
  }
  fclose(f);
  return 0;
}

File: tools/verify/render.js
// Render tools/verify/cases.tsv through the ORIGINAL legacy JS font code and
// print the same digest format as render.c. Any difference between the two
// outputs is an unfaithful port.
const fs = require("fs");
const path = require("path");
const vm = require("vm");

const W = 192, H = 48;
const ROOT = path.join(__dirname, "..", "..");
const px = new Uint8Array(W * H);

// setDot, lifted verbatim from legacy/web/app.js
const ctx = {
  GRID_W: W,
  GRID_H: H,
  setDot(f, x, y, i) {
    if (x < 0 || x >= W || y < 0 || y >= H) return;
    const idx = y * W + x;
    if (i > f[idx]) f[idx] = i;
  },
};
ctx.globalThis = ctx;
vm.createContext(ctx);
vm.runInContext(
  fs.readFileSync(path.join(ROOT, "legacy", "web", "font.js"), "utf8") +
    ";globalThis.__t5 = drawText5; globalThis.__t3 = drawText3;" +
    ";globalThis.__w5 = textWidth5; globalThis.__w3 = textWidth3;",
  ctx
);

function fnv1a(buf) {
  let h = 2166136261 >>> 0;
  for (const b of buf) { h = (h ^ b) >>> 0; h = Math.imul(h, 16777619) >>> 0; }
  return h >>> 0;
}

const casesPath = process.argv[2] || path.join(ROOT, "tools", "verify", "cases.tsv");
for (const raw of fs.readFileSync(casesPath, "utf8").split("\n")) {
  if (!raw || raw.startsWith("/")) continue;
  const line = raw.replace(/\r$/, "");
  const parts = line.split("\t");
  if (parts.length < 7) continue;
  let [name, fn, x, y, inten, scale] = parts;
  const text = parts.slice(6).join("\t");
  x = +x; y = +y; inten = +inten; scale = +scale;

  if (name.startsWith("+")) name = name.slice(1);
  else px.fill(0);

  const adv = fn === "3" ? ctx.__t3(px, x, y, text, inten)
                         : ctx.__t5(px, x, y, text, inten, scale);
  const width = fn === "3" ? ctx.__w3(text) : ctx.__w5(text, scale);

  let nz = 0, x0 = W, y0 = H, x1 = -1, y1 = -1;
  for (let yy = 0; yy < H; yy++)
    for (let xx = 0; xx < W; xx++)
      if (px[yy * W + xx]) {
        nz++;
        if (xx < x0) x0 = xx;
        if (xx > x1) x1 = xx;
        if (yy < y0) y0 = yy;
        if (yy > y1) y1 = yy;
      }

  console.log(
    `${name.padEnd(16)} hash=${fnv1a(px).toString(16).padStart(8, "0")} ` +
    `nz=${String(nz).padEnd(5)} bbox=${x0},${y0},${x1},${y1} adv=${adv} width=${width}`
  );
}
